File: multithreading-multiproc/java/char_freq/CharFreqSeq.java
import java.io.RandomAccessFile;
import java.io.IOException;
import java.nio.charset.StandardCharsets;

class CharFreqSeq {

	static final int N = 128;
    static final int base = 0;
	public static void main(String[] args) throws IOException {

		RandomAccessFile pFile;
		long fileSize;
		char[] buffer;
		String filename;
		int[] freq = new int[N];
		long start, end;

		if(args.length != 1) {
			System.out.println("Usage : CharFreqSeq <file_name>");
			System.exit(1);
			return;
        }
        
        filename = args[0];
        
		try {
			pFile = new RandomAccessFile(filename.toString(), "r");
		} catch(IOException ex) {
			System.out.println("File error");
			System.exit(2);
			return;
		}

		// obtain file size:
		pFile.seek(pFile.length());
		fileSize = pFile.getFilePointer();
		pFile.seek(0);
		System.out.println("file size is " + fileSize);

		byte[] byteArray = new byte[(int)fileSize];
		pFile.readFully(byteArray);

        // allocate memory to contain the file:
        buffer = new String(byteArray, StandardCharsets.UTF_8).toCharArray();

        zeros(freq, N);

		start = System.currentTimeMillis();

		countCharacters(freq, buffer, fileSize);

		end = System.currentTimeMillis();

		displayCount(freq, N);

		System.out.printf("Time spent for counting: %g", (double)(end - start) / 1000);

		pFile.close();
    }

    public static void displayCount(int[] freq, int n) {
		for(int j = 0; j < n; ++j) {
			System.out.println(j + " = " + freq[j]);
		}
	}

    private static void zeros(int[] array, int n){
        for (int j=0; j<n; ++j)
            array[j]=0;
    }
    
    private static void countCharacters(int[] freq, char[] buffer, long fileSize){
        for(int i = 0; i < fileSize; ++i) {
			++freq[buffer[i] - base];
		}
	}
}

File: multithreading-multiproc/python/char_freq/char_freq_seq.py

import sys
from os import SEEK_END
import time
import numpy as np

NUM_OF_ARGS = 2
N = 128
base = 0

def check_args():
    '''
    Function:  check_args 
    ---------------------
    Checks if the number of arguments is correct
    '''
    # if no filename given, exit
    if len(sys.argv) != NUM_OF_ARGS:
        print(f'Usage : {sys.argv[0]} <file_name>\n')
        exit(1)
    return sys.argv[1]

def open_file(mode='r'):
    '''
    Function:  open_file 
    --------------------
    Opens input file size
    mode: method (mode) for opening the file
    '''
    try:
        f = open(filename, mode)
    except Exception as e:
        print(e)
        exit(2)
    return f

def obtain_file_size(f):
    '''
    Function:  obtain_file_size 
    ---------------------------
    Obtains input file's size
    f: input file
    '''
    f.seek(0, SEEK_END)
    file_size = f.tell()
    rewind(f)
    return file_size

def count_characters(freq, buffer, file_size):
    '''
    Function:  count_characters 
    ---------------------------
    Counts the frequency of characters in an char array
    
    freq: the array that will contain each character's frequency
    buffer: the array that contains the characters
    file_size: size of buffer
    '''
    for i in range(file_size):
    	freq[buffer[i] - base] += 1

def rewind(f):
    '''
    Function:  rewind 
    -----------------
    Set the file position to the beginning of the file 
    of the given stream

    f: input file
    '''
    f.seek(0)

if __name__ == '__main__':

    # check arguments, if file name is given return it
    filename = check_args()

    # open file (default mode='r')
    f = open_file()

    # get file size
    file_size = obtain_file_size(f)
    
    # print file size
    print(f'File size is {file_size}\n')

    # frequency vector init with zeros
    freq = np.zeros(N, dtype=np.int32)

    # convert buffer to ascii buffer (numpy array)
    ascii_buff = np.asarray([ord(char) for char in list(f.read(file_size))], dtype=np.int32)

    # start counting
    start = time.time()

    # count characters!
    count_characters(freq, ascii_buff, file_size)

    # done
    end = time.time()

    # print char frequency
    [print(f'{j+base} = {freq[j]}') for j in range(N)]

    # print time of execution
    print(f'Time spent for counting: {end-start}')

    # close file
    f.close()
